### PR TITLE
Object detection example

### PR DESCRIPTION
**Pull request by AngranLi**
This PR consists of 3 parts:
1. The change of xt-training itself to make it work with the dictionary label `y`used in object detection.
2. The new metric of average precision, which is based on the code in https://github.com/rafaelpadilla/Object-Detection-Metrics.
3. The example in Jupyter Notebook. The structure and description is based on Jack's Image Classification example.

Change type:

- [ ] Bug fix
- [x] Feature
- [ ] Documentation

Checklist:

- [x] My code follows the style guidelines of this [project](https://docs.google.com/document/d/1jckZJe0CrWyF-IjxoO2OfBKAyKLC3Yk9Xr7UmOLBETA/edit?usp=sharing)
- [x] I have performed a self-review of my own code
- [x] I have commented my code and added docstrings to all exposed functions and classes
- [x] I have made corresponding changes to the documentation
- [x] Any relevant changes to third party licenses have been updated in the README

### CODE REVIEW

**Review by jwmandeville — DISMISSED** [review]
Looks good and I tested and it works on my system as well.

A few suggestions:
- Add a link to the dataset download page: https://www.cis.upenn.edu/~jshi/ped_html/
- Change the dataset paths and checkpoint paths to variables
- Change the paths to something generic like `path/to/dataset/PennFudanPed` instead of your PC's location

**Comment by AngranLi** [comment]
> Looks good and I tested and it works on my system as well.
> 
> A few suggestions:
> 
> * Add a link to the dataset download page: https://www.cis.upenn.edu/~jshi/ped_html/
> * Change the dataset paths and checkpoint paths to variables
> * Change the paths to something generic like `path/to/dataset/PennFudanPed` instead of your PC's location

Thanks Jack! Just made the changes.

**Review by jwmandeville — DISMISSED** [review]

File: xt_training/runner.py [hunk 1]
```diff
@@ -165,10 +165,31 @@ def __call__(self, loader, mode=None, return_preds=False):
...
-                y_pred = model(x)
-                loss_batch = loss_fn(y_pred, y)
+                        y_tmp = []
+                        for y_i in y:
+                            if isinstance(y_i, dict):
+                                y_i = {k: v.to(device) for k, v in y_i.items()}
+                            else:
+                                y_i.to(device)
+                            y_tmp.append(y_i)
+                        y = y_tmp
```
Comment (timesler) [inline xt_training/runner.py]: Can we add a comment describing what is happening here and what the use case is?

File: xt_training/runner.py [hunk 2]
```diff
@@ -165,10 +165,31 @@ def __call__(self, loader, mode=None, return_preds=False):
...
+                        y_pred = y_pred[-1]
+                except ValueError:
+                    # Object Detection training process requires (x, y) as input.
+                    loss, y_pred = model(x, y)
+                    # During training, y_pred is an empty list with len==0.
+                    # Assign a length to it to make the PooledMean metrics work.
+                    y_pred = [None]
+                    # The Object Detection model already calculated the loss, 
+                    # so loss_fn should be declared to work with it.
+                    loss_batch = loss_fn(y_pred=y_pred, y=loss)
```
Comment (timesler) [inline xt_training/runner.py]: I don't think we should rely on error handling to do non-erroneous control flow.

Also, in general, we shouldn't let the structure of a single object detection model impact the structure of our package so much. I think it would be better to modify the model so that it behaves normally and outputs only the predictions all the time and move the loss calculation to a separate loss function.

File: xt_training/runner.py [hunk 3]
```diff
@@ -165,10 +165,31 @@ def __call__(self, loader, mode=None, return_preds=False):
                     if isinstance(y, torch.Tensor):
                         y = y.to(device)
                     elif isinstance(y, Iterable):
-                        y = [y_i.to(device) for y_i in y]
```
Comment (timesler) [inline xt_training/runner.py]: I'm pretty sure this will error for the data you are expecting. If y_i is a dict, then it won't have a `to` method

Comment (timesler) [reply xt_training/runner.py]: I think a better practice for object detection would be to put the `to` calls inside the `forward` method.

**Review by timesler — CHANGES_REQUESTED** [review]

File: xt_training/runner.py (hunk 1 above)
Comment (AngranLi) [reply xt_training/runner.py]: Comment added.

File: xt_training/runner.py (hunk 2 above)
Comment (AngranLi) [reply xt_training/runner.py]: Yep, I made some modification and discarded the error handling.

File: xt_training/runner.py (hunk 3 above)
Comment (AngranLi) [reply xt_training/runner.py]: Thanks for the suggestion Tim. I wrote an ODInterface based on the SKInterface, and moved the `to` calls inside the `forward` method.

**Comment by AngranLi** [comment]
@timesler The modifications I made:
1. modified `runner.py` so the error handling is discarded
2. add the object detection wrappers so I can move the `to` calls inside the `forward` method.

File: xt_training/metrics.py [hunk 4]
```diff
@@ -197,16 +201,18 @@ def __call__(self, y_pred, y):
         self.latest_value = self.fn(y_pred, y)
         self.latest_num_samples = float(len(y_pred))
         self.num_samples += self.latest_num_samples
-        self.value_sum += self.latest_value.detach() * self.latest_num_samples
+        tmp = self.value_sum + self.latest_value.detach() * self.latest_num_samples
+        # self.value_sum += self.latest_value.detach() * self.latest_num_samples
+        self.value_sum = tmp
```
Comment (timesler) [inline xt_training/metrics.py:206]: Is there any reason for this? I think it's equivalent

File: xt_training/metrics.py [hunk 5]
```diff
@@ -296,6 +302,84 @@ def __init__(self):
...
+        The used library is based on the code by Rafael Padilla.
+    """
+
+    def __init__(self, cls=0, iouthreshold=0.5, method=MethodAveragePrecision.EveryPointInterpolation):
+        """
+            cls: The class that is goint to calculate the average precision.
+            iouthreshold: The IOU threshold to determine whether a detection is true/false positive.
+            method: The interpolation method to calculate the average precision. Can choose between 
+                    every point interpolation or eleven point interpolation.
+        """
```
Comment (timesler) [inline xt_training/metrics.py:316]: This doesn't follow our docstring style. I recommend installing the "Python Docstring Generator" extension in VSCode and configuring it to the correct style. That will make it really easy to add new docstrings with the correct format

File: xt_training/od_lib/BoundingBox.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+            absBB[3],
+            # typeCoordinates=boundingBox.getCoordinatesType(),
+            # imgSize=boundingBox.getImageSize(),
+            bbType=boundingBox.getBBType(),
+            classConfidence=boundingBox.getConfidence(),
+            format=BBFormat.XYWH)
+        return newBoundingBox
+
+
+class BoundingBoxes:
```
Comment (timesler) [inline xt_training/od_lib/BoundingBox.py:165]: If this is intended to be imported by users, we should add a docstring. If not, it should probably be named `_BoundingBoxes` to indicate it's private

**Review by timesler — CHANGES_REQUESTED** [review]
In general, I still think this departs too much from our standard workflow in xt-training, because we are trying to bend over backwards to reproduce a specific tutorial, or to allow the use of the torchvision object detection structure.

In essence, object detection should have no problems conforming to the normal process:
* A dataset returns a tuple of x and y
* The model forward function accepts a single argument (x) and always returns a single variable (y_pred)
* Loss and metric functions accept y and y_pred and return scalar values

The way it's done in torchvision is different from the torchvision classification models but there's not good reason for that as far I know. I think we should build our example using a modified version of the model that doesn't do anything with loss inside the model itself.
